[MRG] Deprecate lsh forest

ldirer opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Fixes #8996 

#### What does this implement/fix? Explain your changes.

Add a deprecation (removed in 0.21) and performance warning for `LSHForest`.

#### Any other comments?

* I could use help on a better message for performance warning ;).
* Should we put the performance warning message in the deprecation message?
Since both are linked it would make sense I think. For now I issued 2 separate warnings.
* I'm puzzled by the `@ignore_warnings` decorator which does not seem to be doing his job in nested function calls: I had to use inline calls to `ignore_warnings` everywhere, which is not super nice.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

ldirer committed:
Add deprecation message and test.

ldirer committed:
Adding performance warning and ignore_warnings in test

ogrisel commented on sklearn/neighbors/approximate.py:
```diff
@@ -217,6 +217,10 @@ def __init__(self, n_estimators=10, radius=1.0, n_candidates=50,
         self.min_hash_match = min_hash_match
         self.radius_cutoff_ratio = radius_cutoff_ratio
 
+        warnings.warn("LSHForest has poor performance.")
```
Please remove that line. Or making an inline comment instead of a dedicated UserWarning.

ldirer replied:
The original issue mentioned 

> LSHForest should be deprecated and scheduled for removal in 0.21. It should also warn about having bad performance

I could make the deprecation warning: 

        LSHForest has poor performance and has been deprecated in 0.19. It will be removed in version 0.21.

ogrisel commented:
Remaining TODO:

- [ ] remove any reference to `LSHForest` in the narrative documentation, API documentation and examples.
- [ ] update `whats_new.rst` to mention the deprecation.

ldirer committed:
Add deprecation to whatsnew and remove LSHForest references from docs.

Removing benchmark for lsh

ldirer commented:
I removed the section on approximate neighbors entirely, but I'm having second thoughts about it. 
I think it's informative (beyond what's implemented in scikit-learn).

If we go ahead and remove it I'll mention in the relevant PRs (that allow ANN, I think PR #8999 does that) that what we removed can be used as a basis for new documentation.

I would have liked to keep some paragraphs but it feels weird to have general considerations that don't relate to any scikit-learn class.

amueller commented on doc/whats_new.rst:
```diff
@@ -427,6 +427,10 @@ API changes summary
 
+   - :class:`neighbors.approximate.LSHForest` has been deprecated and will be
+     removed in 0.21 due to poor performance.
+     :issue:`8996` by `Andreas Müller`_.
```
the by refers to the person coding, not the issue raiser

ogrisel reviewed (approved):
Let's also delete `benchmarks/bench_plot_approximate_neighbors.py` and +1 on my side.

ldirer commented:
Appveyor seems to have randomly failed but otherwise I think it's ready for merge.

amueller commented:
LGTM, thanks
